### DISCUSSION TITLE
Deeplinking at Scale

### DISCUSSION THREAD

**Discussion by deciomontanhani** [MobileNativeFoundation/discussions#144] [General]
Hey y'all,
I would like to start a discussion about how Deeplinking is handled in different companies. I didn't find many articles or videos about deeplinking architecture, troubleshooting, etc.

With that in mind, I would like to know what types of problems, architectures, and solutions do you deal with about deep linking in your work?

Thanks!

**Comment by sberrevoets** [comment]
Deep links are tricky and very difficult to deal with because it applies a web model to clients that weren't built like the web. There's no web server to deal with 404s or redirects when parameters are wrong or missing and the concept of a web page is generally well-defined, but with custom UI like tabs, navigation bars, animations, etc. the concept of a "page" on mobile clients is much more vague.

On the web you jump from 1 page to the next and when navigation happens in line there is no general expectation that linking to a particular page would also navigate _within that page. On mobile it's common to change up only certain elements after the user goes to the next step. There is no built-in infrastructure or tooling to say "this UI can be accessed through this unique identifier/URL".

It gets even more difficult when you add cross-platform into the mix. Any browser can ask for the content behind a URL and then display it, but with mobile apps the UI is built on the device itself (and depending on the app the content can be too) which means you need significant infrastructure to show the right thing in response to following a deep link. Since neither iOS nor Android really provides that layer(AFAIK), it needs to be built twice if you want to support both. More if you want to support more platforms.

Architecturally some of these issues can be solved/mitigated: at Lyft we took some steps to define our own concept of a "page" and gave them unique identifiers so they can be navigated to. The presentation style and error handling is still custom for each deep link, but at least it's bit easier to do the basic things. There's still a lot of edge cases that need to be handled (and in reality aren't handled), but most of the time developers only care about the happy path since few people will enter deep links manually, so it becomes less of an issue.

**Comment by bachand** [comment]
Hey @deciomontanhani . This is a great question. I can provide some context on how we've been approaching deep linking on iOS.

One important consideration at Airbnb is code ownership. Our central Infrastructure teams owns the core deep link service. However, we want each product team to own the URL matching patterns, URL parsing logic, and navigation logic for individual deep links implementations. To achieve this we define each concrete product deep link as a plugin, and those plugins are injected into our core URL handling service. The URL handling service exposes a protocol for a deep link plugin in its interface module (you can learn more about interface modules [here](https://medium.com/airbnb-engineering/designing-for-productivity-in-a-large-scale-ios-application-9376a430a0bf)). In this way our deep linking infrastructure does not have build-time dependencies on modules where deep link implementations are written.

In our system each deep link implementation exposes a set of URL templates that are used for matching concrete URLs.

```swift
// A toy product deep link.

public final class MyDeepLink: DeepLink {
  public let urlTemplates: Set<URLTemplate> = [
    .universalLink(.airbnb, "/foo/:param"),
  ]

  // ...
}
```

In the above code example `.airbnb` refers to the base domain and then the string `"/foo/:param"` matches the path. `:param` indicates that we want to extract the second path component as a named variable called `param` that we can use later in our deep link implementation. Today we use an old version of [DeepLinkKit](https://github.com/button/DeepLinkKit) for URL matching and URL parsing. I expect that we will eventually replace this Objective-C code with a Swift implementation that we write ourselves for a number of reasons.

Allowing the URL templates to be defined in a distributed manner opens up a lot of theoretical problems. For example DeepLinkKit supports regex matching. Ideally it would be great to know if two deep link implementations can match the same URL as that could lead to runtime indeterminism (which deep link implementation do we choose?). However, determining if two regexes overlap is a [complex problem](https://stackoverflow.com/a/1849487). In practice we have found that using regexes is very uncommon. We perform validation at runtime that multiple deep links do not register the exact same URL templates. We also have logic to sort the URL templates so that we will prioritize templates that are concrete over templates that extract parameters (e.g. we will check `/foo/bar` before `/foo/:bar`). 

In our application we have built a modern [routing](https://youtu.be/ray2vMjg2ug?t=661s) system that leverages intent-like objects. Deep link implementations decode URLs to routes. A single deep link implementation can have two different "modes." The first mode that is required for all deep links is being able to handle a URL coming in from the iOS operating system. In this mode the deep link implementation needs to produce one or more routes that the infrastructure will execute sequentially. The deep link can also ask the infrastructure to dismiss visible modals so that the new destination replaces the user's current context. The second mode is optional but is very powerful for server-driven UI experiences. In the second mode the same deep link implementation can decode a URL to a route that they pass back to the caller. Then it's the caller's responsibility to execute the route. In the second mode the caller, who maybe received a URL from the backend API, has full control over how the destination is presented. These two modes are implemented by the same object since code can be shared between the two modes.

When designing a deep link system it's also important to consider the fallback behavior. When a URL does not match any native UI do you fall back to an in-app web view or to the external browser? When a deep link implementation does match a URL but fails to parse the URL who is responsible for the fallback? The answers to questions like these will be specific to your business needs.

Finally, it's important to align between server and client teams on the capabilities and limitations of showing UI from a URL on native. One key difference between native and web is the concept of a navigation stack. On both Android and iOS the user can construct a stack of screens that they will revisit as they go "back" using the platform-specific idiom (the back button on Android, and typically a back arrow or "x" on iOS). You will want to decide if and how this navigation stack can be encoded in URLs, which is more challenging when you consider that App Links and Universal Links are valid web URLs as well.
